core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1039,7 +1039,7 @@ sealed trait ZIO[-R, +E, +A] extends Serializable with ZIOPlatformSpecific[R, E,
    * the specified function to convert the `E` into a `Throwable`.
    */
   final def orDieWith(f: E => Throwable)(implicit ev: CanFail[E]): URIO[R, A] =
-    (self mapError f) catchAll (IO.die(_))
+    self.foldM(e => ZIO.die(f(e)), ZIO.succeedNow)
```

Comment:
@adamgfraser not sure about this one, it looked better with fewer calls.

Reply:
@ahoy-jon Looks good.